capsules/src/net/lowpan_fragment.rs

```diff
@@ -0,0 +1,900 @@
+        match radio.prepare_data_frame(frag_buf,
+                                       self.dst_pan.get(),
+                                       self.dst_mac_addr.get(),
+                                       self.src_pan.get(),
+                                       self.src_mac_addr.get(),
+                                       self.security.get()) {
+            Err(frame) => Err((ReturnCode::FAIL, frame)),
+            Ok(mut frame) => {
+                let dgram_offset = self.dgram_offset.get();
+                let remaining_capacity = frame.remaining_data_capacity() -
+                                         lowpan_frag::FRAGN_HDR_SIZE;
+                // This rounds payload_len down to the nearest multiple of 8 if it
+                // is not the last fragment (per RFC 4944)
+                let remaining_bytes = (self.dgram_size.get() as usize) - dgram_offset;
+                let payload_len = if remaining_bytes > remaining_capacity {
+                    remaining_capacity & !0b111
+                } else {
+                    remaining_bytes
+                };
+
+                // Take the packet temporarily
+                match self.packet.take() {
+                    None => Err((ReturnCode::ENOMEM, frame.into_buf())),
+                    Some(mut packet) => {
+                        let mut frag_header = [0 as u8; lowpan_frag::FRAGN_HDR_SIZE];
+                        set_frag_hdr(self.dgram_size.get(),
+                                     self.dgram_tag.get(),
+                                     dgram_offset,
+                                     &mut frag_header,
+                                     false);
+                        frame.append_payload(&frag_header);
+                        frame.append_payload(&packet[dgram_offset..dgram_offset + payload_len]);
+                        // Replace the packet
+                        self.packet.replace(packet);
+
+                        // Update the offset to be used for the next fragment
+                        self.dgram_offset.set(dgram_offset + payload_len);
+                        let (result, buf) = radio.transmit(frame);
+                        // If buf is returned, then map the error; otherwise, we return success
+                        buf.map(|buf| Err((result, buf))).unwrap_or(Ok(ReturnCode::SUCCESS))
+                    }
+                }
+            }
+        }
+    }
+
+    fn end_transmit(&self, acked: bool, result: ReturnCode) {
+        self.client.get().map(move |client| {
+            // The packet here should always be valid, as we borrow the packet
+            // from the upper layer for the duration of the transmission. It
+            // represents a significant bug if the packet is not there when
+            // transmission completes.
+            let mut packet =
+                self.packet.take().expect("Error: `packet` is None in call to end_transmit.");
```

Comment:
First, the `packet` buffer for the `TxState` struct is statically allocated by the upper layers, and is passed to the `TxState` struct for the duration of the transmission. On the transmit path, ownership of this buffer is passed to `FragState::transmit_packet`, which then passes it to `TxState::init_transmit` (which gives it to the corresponding `TxState` struct). Thus, the `expect` call in `TxState::end_transmit` (line 407) makes sense; if we have lost this buffer, then we are in an inconsistent state, and the upper layer will never receive the buffer back. Since this must be a logic error, we should probably panic.